src/zkp/pifac.rs

```diff
@@ -156,68 +203,86 @@ impl Proof for PiFacProof {
-            let rhs = input.setup_params.scheme().combine(&self.B, &self.Q, &e);
+        let masked_q_commitment_is_valid = {
+            let lhs = input
+                .setup_params
+                .scheme()
+                .reconstruct(&self.q_masked, &self.masked_q_commitment_randomness);
+            let rhs = input.setup_params.scheme().combine(
+                &self.q_mask_commitment,
+                &self.q_commitment,
+                &e,
+            );
             lhs == rhs
         };
-        if !eq_check_2 {
-            error!("eq_check_2 failed");
+        if !masked_q_commitment_is_valid {
+            error!("masked_q_commitment_is_valid failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let eq_check_3 = {
-            let R = input
-                .setup_params
-                .scheme()
-                .reconstruct(&input.N0, self.sigma.as_masked());
-            let lhs = input
+        let modulus_links_provided_factors = {
+            let reconstructed_commitment = input
                 .setup_params
                 .scheme()
-                .reconstruct_with_commitment(&self.Q, &self.z1, &self.v);
-            let rhs = input.setup_params.scheme().combine(&self.T, &R, &e);
+                .reconstruct(&input.modulus, self.link_randomness.as_masked());
+            let lhs = input.setup_params.scheme().reconstruct_with_commitment(
+                &self.q_commitment,
+                &self.p_masked,
+                &self.masked_p_link,
+            );
+            let rhs = input.setup_params.scheme().combine(
+                &self.q_link_commitment,
+                &reconstructed_commitment,
+                &e,
+            );
             lhs == rhs
         };
-        if !eq_check_3 {
-            error!("eq_check_3 failed");
+        if !modulus_links_provided_factors {
+            error!("modulus_links_provided_factors failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let sqrt_N0 = sqrt(&input.N0);
+        let sqrt_modulus = sqrt(&input.modulus);
         // 2^{ELL + EPSILON}
         let two_ell_eps = BigNumber::one() << (ELL + EPSILON);
         // 2^{ELL + EPSILON} * sqrt(N_0)
-        let z_bound = &sqrt_N0 * &two_ell_eps;
-        if self.z1 < -z_bound.clone() || self.z1 > z_bound {
-            error!("self.z1 > z_bound check failed");
+        let z_bound = &sqrt_modulus * &two_ell_eps;
+        if self.p_masked < -z_bound.clone() || self.p_masked > z_bound {
+            error!("p is out of range!");
             return Err(InternalError::ProtocolError);
         }
-        if self.z2 < -z_bound.clone() || self.z2 > z_bound {
-            error!("self.z2 > z_bound check failed");
+        if self.q_masked < -z_bound.clone() || self.q_masked > z_bound {
+            error!("q is out of range!");
             return Err(InternalError::ProtocolError);
         }
```

Comment:
optional nit: You could ditch the clone by negating a reference to the bound, like
```rust
if self.p_masked < -&z_bound || self.p_masked > z_bound {
```
It still looks a big awkward but it avoids the clone at least. Not a big deal though.